### ISSUE TITLE
On our treatment of the output layer

### ISSUE THREAD

**Issue by bpopeters** [OpenNMT/OpenNMT-py#295]
I asked this once before, but the answer I received confused me. For my research, there are some things I want to try out that happen at the output layer. The problem is that it's not completely clear how to do this with OpenNMT. We keep the output layer separate from the rest of the decoder, in the `generator` attribute. We even keep it separate when it's serialized in a checkpoint model. Why is this? I understand the problem has to do with computing the loss efficiently, but why does efficient loss computation require us to act like one of the intrinsic parts of the network is not part of the network?

**Comment by srush** [comment]
I agree we could change this if you proposed something new. 

Let me understand the proposal. The idea would be to have `generator` be part of decoder just like `embeddings`. Instead of outputting its final layer, decoder would just directly output `scores` for each position. This would simplify checkpointing and training code. I agree with this proposal.

The only issue I see is that we do an efficiency hack to avoid keeping the memory of all word positions at once. To get around this problem, we fully parallelize the decoder (maybe in CuDNN), but then have a loop to compute a group of word generator and loss terms. If the word outputs were generated directly, we couldn't do this. 

Do you have any ideas for how to get around this issue? One idea might be to have decoder forward generate a lazy output, that we then compute the loss on one at a time. Not sure I like that. Maybe there is some internal pytorch trick to avoid the memory issue. 

The other point, is that this may not matter for models like yours where you have a small softmax. Maybe that should be the default behavior, and this lazy mode should be a different `forward_lazy` function for decoder.

**Comment by bpopeters** [comment]
I like the `embeddings` analogy. I agree that the challenge will be figuring out how to do this efficiently. I'll see if I can put something together to do this in the next few days. I am admittedly most concerned right now about the issues that affect my own use case, but I'll try to make it efficient for people who work with larger vocabularies as well.

**Comment by srush** [comment]
Okay, I think there is now a nice answer. 

First I will implement this: 
https://github.com/OpenNMT/OpenNMT-py/issues/314

And then I think we can do this PR in a really nice way.
